fpga_interchange/testarch_generators/generate_testarch.py

```diff
@@ -0,0 +1,489 @@
+        site_type = self.device.add_site_type("SLICE")
+
+        # Site pins (with BELs added automatically)
+        site_type.add_pin("L0", Direction.Input)
+        site_type.add_pin("L1", Direction.Input)
+        site_type.add_pin("L2", Direction.Input)
+        site_type.add_pin("L3", Direction.Input)
+        site_type.add_pin("O", Direction.Output)
+
+        site_type.add_pin("R", Direction.Input)
+        site_type.add_pin("C", Direction.Input)
+        site_type.add_pin("D", Direction.Input)
+        site_type.add_pin("Q", Direction.Output)
+
+        # LUT4 BEL
+        bel_lut = site_type.add_bel("LUT", "LUT4", BelCategory.LOGIC)
+        bel_lut.add_pin("I0", Direction.Input)
+        bel_lut.add_pin("I1", Direction.Input)
+        bel_lut.add_pin("I2", Direction.Input)
+        bel_lut.add_pin("I3", Direction.Input)
+        bel_lut.add_pin("O", Direction.Output)
+
+        # DFF BEL
+        bel_ff = site_type.add_bel("FF", "DFF", BelCategory.LOGIC)
+        bel_ff.add_pin("R", Direction.Input)
+        bel_ff.add_pin("C", Direction.Input)
+        bel_ff.add_pin("D", Direction.Input)
+        bel_ff.add_pin("Q", Direction.Output)
+
+        # DFF input mux BEL (routing)
+        bel_mux = site_type.add_bel("FFMUX", "MUX2", BelCategory.ROUTING)
+        bel_mux.add_pin("I0", Direction.Input)
+        bel_mux.add_pin("I1", Direction.Input)
+        bel_mux.add_pin("O", Direction.Output)
+
+        # Site wires
+        w = site_type.add_wire("L0_to_I0", [("L0", "L0"), ("LUT", "I0")])
+        w = site_type.add_wire("L1_to_I1", [("L1", "L1"), ("LUT", "I1")])
+        w = site_type.add_wire("L2_to_I2", [("L2", "L2"), ("LUT", "I2")])
+        w = site_type.add_wire("L3_to_I3", [("L3", "L3"), ("LUT", "I3")])
+
+        w = site_type.add_wire("RST", [("R", "R"), ("FF", "R")])
+        w = site_type.add_wire("CLR", [("C", "C"), ("FF", "C")])
+        w = site_type.add_wire("DIN", [("D", "D"), ("FFMUX", "I1")])
+
+        w = site_type.add_wire("LUT_O")
+        w.connect_to_bel_pin("LUT", "O")
+        w.connect_to_bel_pin("FFMUX", "I0")
+        w.connect_to_bel_pin("O", "O")
+
+        w = site_type.add_wire("MUX_O")
+        w.connect_to_bel_pin("FFMUX", "O")
+        w.connect_to_bel_pin("FF", "D")
+
+        w = site_type.add_wire("FF_OUT", [("FF", "Q"), ("Q", "Q")])
+
+        # Site PIPs
+        site_type.add_pip(("FFMUX", "I0"), ("FFMUX", "O"))
+        site_type.add_pip(("FFMUX", "I1"), ("FFMUX", "O"))
+
+    def make_iob_site_type(self):
+
+        # The site
+        site_type = self.device.add_site_type("IOPAD")
+
+        # Site pins (with BELs added automatically)
+        site_type.add_pin("I", Direction.Output)
+        site_type.add_pin("O", Direction.Input)
+
+        # IPAD bel
+        bel_ib = site_type.add_bel("IB", "IB", BelCategory.LOGIC)
+        bel_ib.add_pin("I", Direction.Output)
+        bel_ib.add_pin("P", Direction.Input)
+
+        bel_ipad = site_type.add_bel("IPAD", "IPAD", BelCategory.LOGIC)
```

Comment:
Not critical for now, but more realistic longer term would be to have one `PAD` bel like Xilinx rather than seperate `IPAD` and `OPAD` bels. This would be useful for testing bidirectional IO away from the complexity of real Xilinx IOBs.

Reply:
Agreed, for now, a `TODO` comment to not lose track of this might be enough.